gctree/dag_filtering.py

```diff
@@ -0,0 +1,507 @@
+        )
+
+    return trimdag
+
+
+def ll_cmcount_dagfuncs(p: np.float64, q: np.float64):
+    """A slower but more numerically stable equivalent to ll_genotype_dagfuncs.
+
+    Args:
+        p, q: branching process parameters
+
+    Returns:
+        A ``historydag.utils.AddFuncDict`` object, which can be unpacked and used
+        as keyword arguments for DAG trimming and weight annotation according to
+        branching process likelihood.
+        Weight format is ``(log likelihood, cmcounts)`` where cmcounts is a FrozenMultiset containing abundance, mutant clade count pairs.
+        To use these functions for DAG trimming, use an optimal function like
+        `lambda l: max(l, key=lambda ll: ll[0])` for clarity, although min or max should work too.
+    """
+
+    funcdict = bp._cmcounter_dagfuncs()
+    cmcount_weight_func = funcdict["edge_weight_func"]
+    cmcount_addfunc = funcdict["accum_func"]
+
+    @lru_cache(maxsize=None)
+    def ll(cmcounter: FrozenMultiset) -> float:
+        if cmcounter:
+            if (0, 1) in cmcounter:
+                cmcounter = cmcounter - {(0, 1)} + {(1, 1)}
+            return bp._lltree(tuple(cmcounter.items()), p, q)[0]
+        else:
+            return 0.0
+
+    def edge_weight_func(n1, n2):
+        st = cmcount_weight_func(n1, n2)
+        return (ll(st), st)
+
+    def accum_func(cmsetlist: List[Tuple[float, FrozenMultiset]]):
+        st = cmcount_addfunc([t[1] for t in cmsetlist])
+        return (ll(st), st)
+
+    return hdag.utils.AddFuncDict(
+        {
+            "start_func": lambda n: (0, FrozenMultiset()),
+            "edge_weight_func": edge_weight_func,
+            "accum_func": accum_func,
+        },
+        names=("LogLikelihood", "cmcounters"),
+    )
+
+
+def ll_genotype_dagfuncs(p: np.float64, q: np.float64):
+    """Return functions for counting tree log likelihood on the history DAG.
+
+    Although these functions are fast, for numerical consistency use
+    :meth:`ll_cmcount_dagfuncs` instead.
+
+    Args:
+        p, q: branching process parameters
+
+    Returns:
+        A ``historydag.utils.AddFuncDict`` object, which can be unpacked and used
+        as keyword arguments for DAG trimming and weight annotation according to
+        branching process likelihood.
+    """
+
+    def edge_weight_ll_genotype(n1: hdag.HistoryDagNode, n2: hdag.HistoryDagNode):
+        """The _ll_genotype weight of the target node, unless it should be
+        collapsed, then 0.
+
+        Expects DAG to have abundances added with
+        :meth:`dag.add_abundances`.
+        """
+        if n2.is_leaf() and n2.label == n1.label:
+            return 0.0
+        else:
+            m = len(n2.clades)
+            # Check if this edge should be collapsed, and reduce mutant descendants
+            if frozenset({n2.label}) in n2.clades:
+                m -= 1
+            return bp.CollapsedTree._ll_genotype(n2.attr["abundance"], m, p, q)[0]
```

Comment:
I'm having trouble understanding this function. Is `n2` the "focal node"? What is the `return 0.0` (i.e. probability 1) condition checking for? If we collapse an edge and decrement `m` don't we need to increment abundance by the label-matched child clade's abundance?

Reply:
The assumptions are that the history DAG is collapsed, and that history DAG nodes have a supplementary label containing abundance information, so that each node with an observed sequence has the observed abundance associated with that sequence. When collapsing the history DAG, leaf-adjacent edges cannot be collapsed, so some trees in the DAG have uncollapsed tips.

We can think of the log likelihood of a CollapsedTree as the sum over all edges in that tree, of the log likelihood associated to the edge's target node's (c, m) pair. (we need to include the root node's likelihood to, of course)
This function is an implementation of that notion of edge weight. `n2` is the edge's target node, and the condition `return 0.0` corresponds to leaf adjacent edges which would be collapsed in a CollapsedTree, but aren't collapsed in the DAG. Because the parent node of such an edge has the same leaf sequence, it's already annotated by the same abundance. Therefore edges targeting that parent node will be assigned the correct abundance and `m` by the conditions in the else statement.

I'm not sure if that's very clear